src/java.base/share/classes/java/util/Formatter.java

```diff
@@ -2856,6 +2850,150 @@ static List<FormatString> parse(String s) {
         return al;
     }
 
+    static final class FormatSpecifierParser {
+        final ArrayList<FormatString> al;
+        final String s;
+        final int max;
+        char first;
+        int start;
+        int off;
+        char c;
+        int argSize;
+        int flagSize;
+        int widthSize;
+
+        FormatSpecifierParser(ArrayList<FormatString> al, char first, int start, String s, int max) {
+            this.al = al;
+
+            this.first = first;
+            this.c = first;
+            this.start = start;
+            this.off = start;
+
+            this.s = s;
+            this.max = max;
+        }
+
+        void reset(char first, int start) {
+            this.first = first;
+            this.c = first;
+            this.start = start;
+            this.off = start;
+
+            argSize = 0;
+            flagSize = 0;
+            widthSize = 0;
+        }
+
+        int parse() {
+            // %[argument_index$][flags][width][.precision][t]conversion
+            // %(\d+\$)?([-#+ 0,(\<]*)?(\d+)?(\.\d+)?([tT])?([a-zA-Z%])
+            int precisionSize = 0;
+
+            parseArgument();
+            parseFlag();
+            parseWidth();
+
+            if (c == '.' && off + 1 < max) {
+                // (\.\d+)?
+                precisionSize = parsePrecision();
+                if (precisionSize == 0) {
+                    return 0;
+                }
+            }
+
+            // ([tT])?([a-zA-Z%])
+            char t = '\0', conversion = '\0';
+            if ((c == 't' || c == 'T') && off + 1 < max) {
+                char c1 = s.charAt(off + 1);
+                if (isConversion(c1)) {
+                    t = c;
+                    conversion = c1;
+                    off += 2;
+                }
+            }
+            if (conversion == '\0' && isConversion(c)) {
+                conversion = c;
+                ++off;
+            }
+
+            if (argSize + flagSize + widthSize + precisionSize + t + conversion != 0) {
+                if (al != null) {
+                    FormatSpecifier formatSpecifier
+                            = new FormatSpecifier(s, start, argSize, flagSize, widthSize, precisionSize, t, conversion);
+                    al.add(formatSpecifier);
+                }
+                return off - start;
+            }
+            return 0;
+        }
+
+        private void parseArgument() {
+            // (\d+\$)?
+            for (int size = 0; off < max; c = s.charAt(++off), size++) {
+                if (!isDigit(c)) {
```

Comment:
You can prepend this check if > '9' to save the table loops happening here under the hood or  just avoid it as a whole (checking the range)? It worth checking this; cause being a digit is indipendent by being ASCII or not, but the table loopup doesn't know it